keps/0024-parameter-enhancement.md

```diff
@@ -0,0 +1,154 @@
+---
+kep-number: 24
+title: Enabling and Disabling features in KUDO Operators
+authors:
+  - "@nfnt"
+owners:
+  - "@nfnt"
+creation-date: 2020-02-21
+last-updated: 2020-02-21
+status: provisional
+---
+
+# Enhanced Operator Parameters
+
+## Table of Contents
+
+- [Enhanced Operator Parameters](#enhanced-operator-parameters)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [Support Optional Parameter Types](#support-optional-parameter-types)
+    - [Provide Parameter Validation](#provide-parameter-validation)
+    - [Notes](#notes)
+  - [Alternatives](#alternatives)
+    - [Convert YAML in string to dictionaries](#convert-yaml-in-string-to-dictionaries)
+    - [Allow Arbitrary Parameter Values](#allow-arbitrary-parameter-values)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+This KEP aims to improve the use-cases that can be covered by parameters. By supporting additional types, templates can access lists and dictionaries. Furthermore, validation can be implemented for known types.
+
+## Motivation
+
+Currently only string parameters are supported by KUDO. This limits the possible use-cases of parameters. Various use cases benefit from having parameters that are lists or dictionaries. While some of this functionality can be implemented in Go templates using [Sprig functions][1], this is unintuitive.
+Having parameter types creates the possibility for parameter validation.
+
+### Goals
+
+- Provide operator developers with more versatile parameters
+- Allow the definition of validation functions for parameters
+
+### Non-Goals
+
+- Change existing parameter definitions in `params.yaml`
+
+## Proposal
+
+### Support Optional Parameter Types
+
+By adding an optional `type` field to the list of parameters, parameter types can be specified. If this field isn't present, it defaults to the `string` type. Other types like `list`, `dict` or `integer` are possible. KUDO's renderer uses the type information to convert the string value to the respective type. For example, for the `dict` type, the input value would be YAML that is unmarshalled to a `map[string]interface{}`. Go's template engine allows to access and iterate list and dictionary items.
```

Comment:
➕to defaults to `string`  type for a strong retro-compatibility 

What about the types of elements of `list` and `dict`? We will always use them as `string`? In that case, should we mention it here

Reply:
I don't think types of elements of `list` and `dict` should be guaranteed to be strings.
Keeping them arbitrary is necessary to be able to support #1221 (and #951).

Reply:
Then we should allow users to specify the structure of those elements to be a custom struct or Integer or String. Or at least specify how it would work. 
cc: @nfnt

Reply:
Elements of `list` and `dicts` can have other types than `string`. Because they will be provided as YAML and YAML supports multiple types.